drivers: add a corrected (and faster) version of CRC16/CCITT

nzmichaelh opened the pull request:
Increases the CRC throughput on the 48 MHz Cortex-M0+ SAMD21
from 239 KiB/s to 2302 KiB/s on a 1 KiB block.

Produces the same output as the bitwise implementation.  See

 https://en.wikipedia.org/wiki/Computation_of_cyclic_redundancy_checks

for how the byte-at-a-time version is calculated.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
Some stats: the function is 44 bytes long using -Os and uses 20 clocks per byte on a Cortex-M0+.

lpereira reviewed (approved):
LGTM: since it's very unlikely that a single application will use two CRC16 implementations, this will in fact result in smaller code when using the CCITT polynomial.

(A quick check confirms this generates the same code when compared with the current implementation and another one with a LUT.)

tbursztyka commented on drivers/crc/crc16_sw.c:
```diff
@@ -34,3 +34,21 @@ u16_t crc16(const u8_t *src, size_t len, u16_t polynomial,
+
+	for (i = 0; i < len; i++) {
+		/*
+		 * See https://en.wikipedia.org/wiki/
```
remove that comment, you already clearly pint pointed the source in your commit message.

tbursztyka commented on drivers/crc/crc16_sw.c:
```diff
@@ -34,3 +34,21 @@ u16_t crc16(const u8_t *src, size_t len, u16_t polynomial,
+
+		crc = (crc << 8) ^ t ^ (t << 5) ^ (t << 12);
+	}
+	return crc;
```
break a line before return crc

nzmichaelh replied:
Done.

tbursztyka reviewed (changes requested):
I think it would be good to have the same optimization for crc16_ansi then.

polynomial version should be kept for non contiguous buffers (that was the point of initial_value/pad parameters afaik)

pfl commented on include/crc16.h:
```diff
@@ -44,10 +44,7 @@ u16_t crc16(const u8_t *src, size_t len, u16_t polynomial,
-{
-	return crc16(src, len, 0x1021, 0xffff, true);
-}
+u16_t crc16_ccitt(const u8_t *src, size_t len);
```
Speeding up the computation is awsome, but I have a few comments:
1) why is the improvement made only for the ccitt variant and not the generic one? All other variants will now be slower without any good reason.
2) test cases must be added, else the next person (or even this implementation) cannot be known to compute correct values
3) this implementation works only on continuous buffers, which is not the case usually. This is why bool pad was introduced later on.

nzmichaelh replied:
Hi @pfl.  Thanks for the review!

re: optimisation, I've added test vectors and a byte-at-a-time version for crc16_ansi.

re: test cases, there's already vectors in tests/unit/drivers/crc/main.c.  I've added vectors for crc16_ansi.

re: pad, the bitwise version also only supports a single, contiguous buffer.  'pad' is something different - it causes an extra two bytes of zeros to be fed through the CRC function as required by the CCITT spec.  The bytewise version achieves the same by using a different seed.  See http://srecord.sourceforge.net/crc16-ccitt.html for more.

nzmichaelh replied:
Hi @tbursztyka see my reply to @pfl above.

pfl replied:
Thanks for the tests!
For 1) above I actually meant that there would be one function that computes the CRC, and then the variants would call this function with their initial values and polynomial. But something smart needs to be done along the way for XOR, since it now is not as trivial as before when it was done bit by bit.
About the pad value: yes, the input buffer is continuous. But having the pad added or not allowed the computation to continue thus enabling the computation to be done over N continuous buffers, see commit 5d8eadd7b1aa19b1b33ce88f2a06ca57f6222c10 where it was introduced. This is needed by some, so that should still be supported.

nzmichaelh replied:
Ah, sorry, in my earlier reply I got 'bitwise' and 'bytewise' mixed up in my head.

re: crc16, sorry, I don't know a way of doing a fast, runtime configured crc16.  The bytewise version is 10x faster as the loop is unrolled and known at compile time.  The best options I can think of are:

1. Change crc16 to call the specialised ones when the parameters are the right values.  This would increase the code size as all versions would get linked in.

2. Switch to a table driven method as the table can be filled in on first use.  This would need a cache and 256 x 2 bytes of RAM for the table itself.

re: the use case, you can chain the bytewise versions by changing the API to allow the initial CRC to be passed in.  Something like:

`u16_t crc16_ccitt_seeded(u16_t initial_crc, void *buf, int len, bool final)`

I'm not sure that 'final' is needed - I'm pretty sure that if you use 0x1d01 as a seed instead of 0xffff then you can skip the final 16 bits of zeros.

nzmichaelh replied:
I looked further, and the only user of crc16 is the nffs which is originally from mynewt.  They use the chained CRC method:

https://github.com/apache/mynewt-core/blob/master/fs/nffs/src/nffs_crc.c#L110

with the following API:

https://github.com/apache/mynewt-core/blob/master/util/crc/src/crc16.c#L69

msink commented:
Implementation used in all my projects:
```
/*
 * Optimized CRC-CCITT calculation
 * Polynomial: x^16 + x^12 + x^5 + 1 (0x8408)
 */
u16 crc_ccitt_update(u16 crc, u8 data)
{
        data ^= (u8)crc;
        data ^= data << 4;
        return (((u16)data << 8) | (u8)(crc >> 8)) ^
                (u8)(data >> 4) ^ ((u16)data << 3);
}

/**
 * crc_ccitt - recompute the CRC for the data buffer
 * @crc: previous CRC value
 * @buffer: data pointer
 * @len: number of bytes in the buffer
 */
u16 crc_ccitt(u16 crc, u8 const *buffer, size_t len)
{
        while (len--)
                crc = crc_ccitt_update(crc, *buffer++);
        return crc;
}
```

It uses LSBF/reverse bitorder, but result and speed should be the same.

nzmichaelh commented:
As an aside, I've been looking into the different CRC16-CCITT implementations out there.  The short story is that every project is different :)

* Zephyr uses CRC-16/AUG-CCITT which is equivalent to CRC-16/CCITT-FALSE with two zero bytes prepended.  The effective seed is `0x1d0f` and test vector output is `0xe5cc`
* Mynewt uses CRC-16/XMODEM.  The seed is 0 and test vector output is `0x31c3`
* Contiki uses CRC-16/CCITT.  The seed is 0, input and output are bit reversed, and test vector output is `0x2189`
* Linux uses CRC-16/MCRF4XX.   The seed is `0xffff`, input and output are bit reversed, and the test vector output is `0x6f91`
* The SAMD1 uses CRC-16/CCITT as per Contiki

See http://reveng.sourceforge.net/crc-catalogue/16.htm for a catalogue.

nzmichaelh committed:
crc: add a faster, chainable version of CRC16/CCITT.

The existing version of crc16_ccitt() is actually CRC-16/AUG-CCITT and
gives different results to Linux, Contiki, and the CRC unit in the
SAM0 SOC.  This version matches Linux.

Note that this is an incompatible API change.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
Hi @pfl I looked further and things get... interesting.  crc16_ccitt() isn't actually CRC16/CCITT - it's in the same family but generates different results to the similarly named functions used in Linux and Contiki.

WDYT about the updated patch?  It leaves the existing CRC routines as-is to preserve compatibility and adds a new function.  The new function matches the Linux API [1] and allows chaining across non-contigious blocks in the same way that the Mynewt version of nffs does [2].

For context, I'm hacking on a project using the SAMD21 and I'd like to use the built-in CRC unit in the future as it's another 4x faster than software.  The CRC unit can only do CRC-32 and CRC-16/CCITT.

[1] https://github.com/spotify/linux/blob/6eb782fc88d11b9f40f3d1d714531f22c57b39f9/lib/crc-ccitt.c#L60

[2] https://github.com/apache/mynewt-core/blob/c940449834cd4f4fa5ee57ec4480463d056b2260/fs/nffs/src/nffs_crc.c#L46

msink commented:
Don't know what variant is mine (above), but it used in ISDN hdlc frames like this:
```
    u16 crc = ~crc_ccitt(0xffff, hdlc_buf, len);
    hdlc_buf[len + 0] = crc;
    hdlc_buf[len + 1] = crc >> 8;
```
```
    if (len > 2 &&  crc_ccitt(0xffff, hdlc_buf, len) == 0xf0b8)
```

nzmichaelh commented:
Hi @msink yip, that's the PPP / X.25 version:  0xffff is the seed, the ~ is equivalent to xoring the output with 0xffff, and 0xf0b8 is the residual.

tbursztyka commented on drivers/crc/crc_ccitt_sw.c:
```diff
@@ -0,0 +1,20 @@
+
+#include <crc16.h>
+
+u16_t crc_ccitt(u16_t seed, const u8_t *src, size_t len)
```
Move this into drivers/crc/crc16_sw.c  : it is a software way of doing a 16bits crc.

nzmichaelh replied:
Hi @tbursztyka, thanks for the review.  Done.

Note that this may need refactoring later - I want to add support for the hardware CRC-16/CCITT unit in the SAMD21 but it'll still need software fallback for crc16_ansi etc.

tbursztyka commented on include/crc16.h:
```diff
@@ -74,6 +74,30 @@ static inline u16_t crc16_ansi(const u8_t *src, size_t len)
+ *
+ * @return The computed CRC16 value
+ */
+u16_t crc_ccitt(u16_t seed, const u8_t *src, size_t len);
```
Keep the name crc16_ccitt, replace the old one directly. Do we really need to keep 2 versions?

nzmichaelh replied:
SGTM.  To confirm, this will change the API as the new version adds a third parameter to support non-contiguous buffers for nffs or, say, chained net_bufs.  Is that OK?

Note Linux also uses 'crc_ccitt()' - see linux/include/crc-ccitt.h.

tbursztyka commented on tests/unit/drivers/crc/main.c:
```diff
@@ -30,11 +31,51 @@ void test_crc16_ansi(void)
 	zassert(crc16_ansi(test2, sizeof(test2)) == 0x9ecf, "pass", "fail");
 }
 
+void test_crc_ccitt(void)
```
keep the 16 in crc16.

nzmichaelh replied:
Will do based on the answer to the previous comment.

nashif commented:
@nzmichaelh  this PR seems to be stale, can you please rebase?

nzmichaelh committed:
crc: add test vectors for the ANSI version.

The test cases match
http://reveng.sourceforge.net/crc-catalogue/16.htm for
CRC-16/DDS-110.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
@nashif rebased.  If this looks good then I'll update the WIP SDHC SPI driver to use these routines.

nzmichaelh committed:
crc: add crc16_itu_t which is used in the MMC protocol.

This is the MSB first version of crc16_ccitt.  Use the same name as in
Linux.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
I've rebased and added crc16_itu_t which is the MSB first version of the same CRC and is the one used in the MMC protocol.

nashif commented:
@pfl @tbursztyka  please re-review!

pfl commented:
Looks good to me.
